src/layers/SchematicLayer.ts

```diff
@@ -428,6 +507,288 @@ export class SchematicLayer<T extends SchematicData> extends PixiLayer<T> {
     return this.cementPlugTextureCache;
   }
 
+  private createCementPlugShape(plug: CementPlug, casings: Casing[], holes: HoleSize[]): ComplexRopeSegment[] {
+    const { exaggerationFactor } = this.options as SchematicLayerOptions<T>;
+    return createComplexRopeSegmentsForCementPlug(plug, casings, holes, exaggerationFactor, this.getZFactorScaledPathForPoints);
+  }
+
+  private createPerforationTexture(perforation: Perforation, perfShapes: PerforationShape[], otherPerforations: Perforation[]): Texture {
+    if (this.perforationTextureCache) {
+      return this.perforationTextureCache;
+    }
+
+    const { cementOptions } = this.options as SchematicLayerOptions<T>;
+    const canvas = document.createElement('canvas');
+
+    const size = DEFAULT_TEXTURE_SIZE * cementOptions.scalingFactor;
+
+    canvas.width = size / 2;
+    canvas.height = size;
+    const canvasCtx = canvas.getContext('2d');
+
+    canvasCtx.fillStyle = perforationColors.red;
+
+    const spikeWidth = 25;
+    const amountOfSpikes = canvas.width / spikeWidth;
+    const fracLineHalfWidth = 10;
+    const packingOpacity = 0.5;
+
+    // how to start the spike at the right diameter position if it has multiple perforationShapes with different diameters?
+    // maybe good enough to just render it behind casings or holes?
+    // @ooystein
+
+    // https://app.zenhub.com/workspaces/wellx-5f89a0db386bba0014989b28/issues/equinor/wellx-designer/222
+    foldPerforationSubKind(
```

Comment:
Can you move the creation of the cement texture to a new function in [`schematicShapeGenerator`](https://github.com/equinor/esv-intersection/blob/master/src/datautils/schematicShapeGenerator.ts)?
Splitting the logic into multiple functions per subkind could be nice as well. 😄 
The comments are great 🤩 Have a second pass on them and see if anything can be removed, or if there's outstanding questions.